x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/BinarySpatialFunction.java

```diff
@@ -80,6 +81,29 @@ public void writeTo(StreamOutput out) throws IOException {
         // The CRS type is re-resolved from the combination of left and right fields, and also not necessary to serialize
     }
 
+    /**
+     * Mark the function as expecting the specified fields to arrive as doc-values.
+     */
+    public abstract BinarySpatialFunction withDocValues(boolean foundLeft, boolean foundRight);
```

Comment:
This code pre-exists this PR by about half a year. It just got moved up to the parent class, because there was a bug with ST_DISTANCE and centroid aggregations. Having said that, I also cannot see how to move this out of the plan nodes, because this is knowledge generated during planning that is then used to influence field extraction later. The knowledge of whether we use doc values is determined by the planning rule `SpatialDocValuesExtraction` which needs to look at a tree of physical plan nodes to determine if the right combination exists to take this decision, and then mark several nodes in the plan with this knowledge, notably the centroid aggregation itself, and any spatial functions that might be affected by this (any functions called before the centroid aggregation).